src/main/resources/lib/portal.ts

```diff
@@ -1,16 +1,70 @@
-import { query, Site } from "/lib/xp/content";
-import { forceArray } from "./utils";
+import { query, type Site, get } from "/lib/xp/content";
+import { forceArray, buildBaseContext } from "./utils";
+import { list, type Repository } from "/lib/xp/repo";
+import { run } from "/lib/xp/context";
+import type { SiteConfig } from "/site/index";
+import { EnonicEventDataNode } from "/lib/xp/event";
 
-export function getSiteInCron<Config extends object>(): Site<Config> {
+export interface RepoSiteAppConfig {
+  repoId: string,
+  siteName: string,
+  appConfig: SiteConfig,
+}
+
+function getSites<Config extends object>(): Site<Config>[] {
   return query<Site<Config>>({
-    count: 1,
     query: `_path LIKE '/content/*' AND data.siteConfig.applicationKey = '${app.name}'`,
-    contentTypes: ["portal:site"],
-  }).hits[0];
+    contentTypes: ["portal:site"]
+  }).hits;
+}
+
+const runInDraftRepoContext = (callback: () => void, repositoryId: string, params: []) => {
+  if (!repositoryId) {
+    repositoryId = "com.enonic.cms.default"; // Best guess... This shouldn't happen!
+  }
+
+  return run(buildBaseContext(repositoryId), () => callback(...params));
+};
+
+export function getAllSiteConfigsInCron(): RepoSiteAppConfig[] {
+  const siteConfigs: RepoSiteAppConfig[] = [];
+  const repolist: Repository[] = list();
+  const filteredRepolist = repolist.filter( (repository) => repository.branches.indexOf("draft") >= 0 )
+    .map( (repository) => repository.id );
+
+  filteredRepolist.forEach((repositoryId) => {
```

Comment:
Can you see if you can refactor this with `map` instead of `forEach`. (If it's a lot of work, this is not super important for me).

I usually only use `forEach` in frontend code, where I need to do mutations to the global state of the page.

For backend code, we usually think of what we are writing as "data transformation pipelines". Don't be afraid to extract each step in the `map` / `filter` / `flatMap` pipeline into separate funcitons.

---

Here is the implementation of `flatMap` I usually use:

```typescript
export function flatMap<A, B>(arr: Array<A>, f: (a: A) => Array<B>): Array<B> {
  return arr.reduce<B[]>((acc, val) => acc.concat(f(val)), []);
}
```